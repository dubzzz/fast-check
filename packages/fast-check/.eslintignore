bundle/
node_modules/
lib/
lib-*/
perf/
jest*.js
jest*.cjs