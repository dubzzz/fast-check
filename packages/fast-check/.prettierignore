bundle/
coverage/
docs/
documentation/Arbitraries.md
lib/
lib-*/