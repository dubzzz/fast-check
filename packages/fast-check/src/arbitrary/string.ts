import type { Arbitrary } from '../check/arbitrary/definition/Arbitrary';
import type { ArrayConstraintsInternal } from './array';
import { array } from './array';
import type { StringSharedConstraints } from './_shared/StringSharedConstraints';
import { createSlicesForString } from './_internals/helpers/SlicesForStringBuilder';
import { stringUnit } from './_internals/StringUnitArbitrary';
import { patternsToStringMapper, patternsToStringUnmapperFor } from './_internals/mappers/PatternsToString';
export type { StringSharedConstraints } from './_shared/StringSharedConstraints';

/**
 * Constraints to be applied on arbitrary {@link string}
 * @remarks Since 3.22.0
 * @public
 */
export type StringConstraints = StringSharedConstraints & {
  /**
   * A string results from the join between several unitary strings produced by the Arbitrary instance defined by `unit`.
   * The `minLength` and `maxLength` refers to the number of these units composing the string. In other words it does not have to be confound with `.length` on an instance of string.
   *
   * A unit can either be a fully custom Arbitrary or one of the pre-defined options:
   * - `'grapheme'` - Any printable grapheme as defined by the Unicode standard. This unit includes graphemes that may:
   *   - Span multiple code points (e.g., `'\u{0061}\u{0300}'`)
   *   - Consist of multiple characters (e.g., `'\u{1f431}'`)
   *   - Include non-European and non-ASCII characters.
   *   - **Note:** Graphemes produced by this unit are designed to remain visually distinct when joined together.
   * - `'grapheme-composite'` - Any printable grapheme limited to a single code point. This option produces graphemes limited to a single code point.
   *   - **Note:** Graphemes produced by this unit are designed to remain visually distinct when joined together.
   * - `'grapheme-ascii'` - Any printable ASCII character.
   * - `'binary'` - Any possible code point (except half surrogate pairs), regardless of how it may combine with subsequent code points in the produced string. This unit produces a single code point within the full Unicode range (0000-10FFFF).
   * - `'binary-ascii'` - Any possible ASCII character, including control characters. This unit produces any code point in the range 0000-00FF.
   *
   * @defaultValue 'grapheme-ascii'
   * @remarks Since 3.22.0
   */
  unit?: 'grapheme' | 'grapheme-composite' | 'grapheme-ascii' | 'binary' | 'binary-ascii' | Arbitrary<string>;
};

/** @internal */
function extractUnitArbitrary(constraints: Pick<StringConstraints, 'unit'>): Arbitrary<string> {
  if (typeof constraints.unit === 'object') {
    return constraints.unit;
  }
  switch (constraints.unit) {
    case 'grapheme':
      return stringUnit('grapheme', 'full');
    case 'grapheme-composite':
      return stringUnit('composite', 'full');
    case 'grapheme-ascii':
    case undefined:
      return stringUnit('grapheme', 'ascii');
    case 'binary':
      return stringUnit('binary', 'full');
    case 'binary-ascii':
      return stringUnit('binary', 'ascii');
  }
}

/**
 * For strings of {@link char}
 *
 * @param constraints - Constraints to apply when building instances (since 2.4.0)
 *
 * @remarks Since 0.0.1
 * @public
 */
export function string(constraints: StringConstraints = {}): Arbitrary<string> {
  const charArbitrary = extractUnitArbitrary(constraints);
  const unmapper = patternsToStringUnmapperFor(charArbitrary, constraints);
  const experimentalCustomSlices = createSlicesForString(charArbitrary, constraints);
  const enrichedConstraints: ArrayConstraintsInternal<string> = { ...constraints, experimentalCustomSlices };
  return array(charArbitrary, enrichedConstraints).map(patternsToStringMapper, unmapper);
}
