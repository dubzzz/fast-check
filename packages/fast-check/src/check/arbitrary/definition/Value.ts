import type { WithCloneMethod } from '../../symbols';
import { cloneMethod, hasCloneMethod } from '../../symbols';

const safeObjectDefineProperty = Object.defineProperty;

/**
 * A `Value<T, TShrink = T>` holds an internal value of type `T`
 * and its associated context
 *
 * @remarks Since 3.0.0 (previously called `NextValue` in 2.15.0)
 * @public
 */
export class Value<T> {
  /**
   * State storing the result of hasCloneMethod
   * If `true` the value will be cloned each time it gets accessed
   * @remarks Since 2.15.0
   */
  readonly hasToBeCloned: boolean;
  /**
   * Flag indicating whether or not the this.value has already been called once
   * If so, the underlying will be cloned
   * Only set when hasToBeCloned = true
   */
  private readOnce: boolean;
  /**
   * Safe value of the shrinkable
   * Depending on `hasToBeCloned` it will either be `value_` or a clone of it
   * @remarks Since 2.15.0
   */
  readonly value!: T;
  /**
   * Internal value of the shrinkable
   * @remarks Since 2.15.0
   */
  readonly value_: T;
  /**
   * Context for the generated value
   * TODO - Do we want to clone it too?
   * @remarks 2.15.0
   */
  readonly context: unknown;

  /**
   * @param value_ - Internal value of the shrinkable
   * @param context - Context associated to the generated value (useful for shrink)
   * @param customGetValue - Limited to internal usages (to ease migration to next), it will be removed on next major
   */
  constructor(value_: T, context: unknown, customGetValue: (() => T) | undefined = undefined) {
    this.value_ = value_;
    this.context = context;
    this.hasToBeCloned = customGetValue !== undefined || hasCloneMethod(value_);
    this.readOnce = false;
    if (this.hasToBeCloned) {
      safeObjectDefineProperty(this, 'value', { get: customGetValue !== undefined ? customGetValue : this.getValue });
    } else {
      this.value = value_;
    }
  }

  /** @internal */
  private getValue() {
    if (this.hasToBeCloned) {
      if (!this.readOnce) {
        this.readOnce = true;
        return this.value_;
      }
      return (this.value_ as unknown as WithCloneMethod<T>)[cloneMethod]();
    }
    return this.value_;
  }
}
