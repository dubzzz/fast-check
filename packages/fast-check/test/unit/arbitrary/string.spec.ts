import { describe, it, expect } from 'vitest';
import * as fc from 'fast-check';
import { string } from '../../../src/arbitrary/string';
import type { StringConstraints } from '../../../src/arbitrary/string';

import { Value } from '../../../src/check/arbitrary/definition/Value';
import {
  assertProduceValuesShrinkableWithoutContext,
  assertShrinkProducesSameValueWithoutInitialContext,
  assertProduceCorrectValues,
  assertProduceSameValueGivenSameSeed,
} from './__test-helpers__/ArbitraryAssertions';
import { buildShrinkTree, renderTree } from './__test-helpers__/ShrinkTree';
import { Arbitrary } from '../../../src/check/arbitrary/definition/Arbitrary';
import type { Random } from '../../../src/random/generator/Random';
import { Stream } from '../../../src/stream/Stream';

describe('string (integration)', () => {
  type Extra = StringConstraints;
  const extraParameters: fc.Arbitrary<Extra> = fc
    .tuple(
      fc.nat({ max: 5 }),
      fc.nat({ max: 30 }),
      fc.boolean(),
      fc.boolean(),
      fc.constantFrom<StringConstraints['unit']>(
        undefined,
        'grapheme',
        'grapheme-composite',
        'grapheme-ascii',
        'binary',
        'binary-ascii',
        new PatternsArbitrary(['123', 'abc', '!']),
      ),
    )
    .map(([min, gap, withMin, withMax, unit]) => ({
      minLength: withMin ? min : undefined,
      maxLength: withMax ? min + gap : undefined,
      unit,
    }));

  const measureLengthFor = (value: string, extra: Extra): number => {
    if (typeof extra.unit === 'object') {
      return [...value.matchAll(/(abc|123|!)/g)].length;
    }
    switch (extra.unit) {
      case 'grapheme':
      case 'grapheme-composite':
      case 'grapheme-ascii':
      case undefined: {
        // @ts-expect-error Not available with our current preset for TypeScript
        const segmenter = new Intl.Segmenter();
        return [...segmenter.segment(value)].length;
      }
      case 'binary':
      case 'binary-ascii':
        return [...value].length;
    }
  };

  const assertCorrectForUnit = (value: string, extra: Extra): void => {
    if (typeof extra.unit === 'object') {
      expect(value).toMatch(/^(abc|123|!)*$/); // based on PatternsArbitrary
      return;
    }
    switch (extra.unit) {
      case 'grapheme':
        return;
      case 'grapheme-composite': {
        // @ts-expect-error Not available with our current preset for TypeScript
        const segmenter = new Intl.Segmenter();
        expect([...segmenter.segment(value)].map((s) => s.segment)).toEqual([...value]); // assert: grapheme equivalent to code-point
        return;
      }
      case 'grapheme-ascii':
      case undefined:
        expect([...value]).toEqual(value.split('')); // assert: code point equivalent to code unit aka char
        for (const c of value) {
          expect(c.charCodeAt(0)).toBeGreaterThanOrEqual(0x20);
          expect(c.charCodeAt(0)).toBeLessThanOrEqual(0x7e);
        }
        return;
      case 'binary':
        return;
      case 'binary-ascii':
        expect([...value]).toEqual(value.split('')); // assert: code point equivalent to code unit aka char
        for (const c of value) {
          expect(c.charCodeAt(0)).toBeGreaterThanOrEqual(0x00);
          expect(c.charCodeAt(0)).toBeLessThanOrEqual(0xff);
        }
        return;
    }
  };

  const isCorrect = (value: string, extra: Extra) => {
    const measuredLength = measureLengthFor(value, extra);
    if (extra.minLength !== undefined) {
      expect(measuredLength).toBeGreaterThanOrEqual(extra.minLength);
    }
    if (extra.maxLength !== undefined) {
      expect(measuredLength).toBeLessThanOrEqual(extra.maxLength);
    }
    assertCorrectForUnit(value, extra);
  };

  const stringBuilder = (extra: Extra) => string(extra);

  it('should produce the same values given the same seed', () => {
    assertProduceSameValueGivenSameSeed(stringBuilder, { extraParameters });
  });

  it('should only produce correct values', () => {
    assertProduceCorrectValues(stringBuilder, isCorrect, { extraParameters });
  });

  it('should produce values seen as shrinkable without any context', () => {
    assertProduceValuesShrinkableWithoutContext(stringBuilder, { extraParameters });
  });

  it('should be able to shrink to the same values without initial context', () => {
    assertShrinkProducesSameValueWithoutInitialContext(stringBuilder, { extraParameters });
  });

  const patterns = new PatternsArbitrary(['123', 'abc', '!']);
  const allExcept = (exclusionList: StringConstraints['unit'][]): StringConstraints['unit'][] => {
    const all: StringConstraints['unit'][] = [
      'grapheme' as const,
      'grapheme-composite' as const,
      'grapheme-ascii' as const,
      'binary' as const,
      'binary-ascii' as const,
      patterns,
    ];
    return all.filter((e) => !exclusionList.includes(e));
  };
  describe.each([
    { unit: undefined },
    { unit: 'grapheme' as const },
    { unit: 'grapheme-composite' as const },
    { unit: 'grapheme-ascii' as const },
    { unit: 'binary' as const },
    { unit: 'binary-ascii' as const },
    { unit: patterns },
  ])('based on unit=$unit', ({ unit }) => {
    it.each<{ source: string; constraints: Omit<StringConstraints, 'unit'>; supportedBy: StringConstraints['unit'][] }>(
      [
        // Empty string is valid for any unit as long as there is no minLength>0...
        { source: '', constraints: {}, supportedBy: allExcept([]) },
        // ...If we ask for minLength=1, it becomes invalid
        { source: '', constraints: { minLength: 1 }, supportedBy: [] },
        // Except patternsArbitrary all units are OK with the strings below as they don't break any constraints (neither content nor length)...
        { source: 'azerty', constraints: {}, supportedBy: allExcept([patterns]) },
        { source: 'ah! ah!', constraints: {}, supportedBy: allExcept([patterns]) },
        // ...just in case we play with the default implicit maxLength (=maxLengthFromMinLength(0)=10)...
        { source: '0123456789', constraints: {}, supportedBy: allExcept([patterns]) },
        // ...and just above it as when not provided we should still accept to shrink for any lengths even if we do not generate them.
        { source: '0123456789'.repeat(10), constraints: {}, supportedBy: allExcept([patterns]) },
        // On its own patternsArbitrary is also able to generate things but its language is way more limited (with our configuration)...
        { source: 'abc!123!', constraints: {}, supportedBy: allExcept([]) },
        // ...it does not count the length the same way as such 'abc', '123' or '!' are all considered to be a length 1.
        // So it is the only one to accept 'abc!123!' when asked for a maxLength of 4.
        { source: 'abc!123!', constraints: { maxLength: 4 }, supportedBy: [patterns] },
        // Regarding not supported characters we have many examples even for others:
        // - ascii versions do not support code-points out-side of the ascii range
        { source: 'a\u{1f431}b', constraints: {}, supportedBy: ['binary', 'grapheme', 'grapheme-composite'] },
        // - composite versions do not support decomposed code-points composing themselves with others (ascii neither)
        { source: '\u{0061}\u{0300}', constraints: {}, supportedBy: ['binary', 'grapheme'] },
        // - no one support half surrogate pairs (high or low)
        { source: '\u{d800}', constraints: {}, supportedBy: [] },
        { source: '\u{dfff}', constraints: {}, supportedBy: [] },
        // Then we may discard entries as they break the length rules:
        // - not large enough
        { source: 'ab', constraints: { minLength: 3 }, supportedBy: [] },
        { source: 'abc', constraints: { minLength: 4 }, supportedBy: [] },
        // - too large
        { source: 'abcd', constraints: { maxLength: 3 }, supportedBy: [] },
        { source: 'abcabcabc', constraints: { maxLength: 2 }, supportedBy: [] },
        { source: '0123456789', constraints: { maxLength: 9 }, supportedBy: [] },
        { source: '0123456789'.repeat(10), constraints: { maxLength: 10 }, supportedBy: [] },
        // But counting characters is different from one unit to another, not only just for patternsArbitrary.
        // For binary: '\u{0061}\u{0300}' is ['\u{0061}', '\u{0300}'] -> length 2
        // For grapheme: '\u{0061}\u{0300}' is ['\u{0061}\u{0300}']   -> length 1
        { source: '\u{0061}\u{0300}', constraints: { minLength: 2 }, supportedBy: ['binary'] },
        { source: '\u{0061}\u{0300}', constraints: { maxLength: 1 }, supportedBy: ['grapheme'] },
      ],
    )(
      'should be able to generate $source with fc.string($constraints) for $supportedBy',
      ({ source, constraints, supportedBy }) => {
        // Arrange / Act
        const arb = string({ ...constraints, unit });
        const out = arb.canShrinkWithoutContext(source);

        // Assert
        expect(out).toBe(supportedBy.includes(unit ?? 'grapheme-ascii'));
      },
    );
  });

  it.each([
    { unit: undefined, rawValue: 'Hey!' },
    { unit: undefined, rawValue: '0'.repeat(50) /* longer than default maxGeneratedLength but ok for shrink */ },
    { unit: undefined, rawValue: 'abc!123!' },
    { unit: patterns, rawValue: 'abc!123!' },
  ])('should be able to shrink $rawValue with fc.string({$unit})', ({ rawValue, unit }) => {
    // Arrange
    const arb = string({ unit });
    const value = new Value(rawValue, undefined);

    // Act
    const renderedTree = renderTree(buildShrinkTree(arb, value, { numItems: 100 })).join('\n');

    // Assert
    expect(arb.canShrinkWithoutContext(rawValue)).toBe(true);
    expect(renderedTree).toMatchSnapshot();
  });
});

// Helpers

class PatternsArbitrary extends Arbitrary<string> {
  constructor(readonly patterns: string[]) {
    super();
  }
  generate(mrng: Random): Value<string> {
    return new Value(this.patterns[mrng.nextInt(0, this.patterns.length - 1)], undefined);
  }
  canShrinkWithoutContext(value: unknown): value is string {
    if (typeof value !== 'string') return false;
    return this.patterns.includes(value);
  }
  shrink(value: string): Stream<Value<string>> {
    const patternIndex = this.patterns.indexOf(value);
    if (patternIndex <= 0) {
      return Stream.nil();
    }
    return Stream.of(new Value(this.patterns[0], undefined));
  }
}
