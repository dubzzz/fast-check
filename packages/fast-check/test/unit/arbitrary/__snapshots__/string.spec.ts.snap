// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`string (integration) > should be able to shrink '0000000000000000000000000000000000000…' with fc.string({undefined}) 1`] = `
""00000000000000000000000000000000000000000000000000"
├> ""
├> "0000000000000000000000000"
|  ├> "0000000000000"
|  |  ├> "0000000"
|  |  |  ├> "0000"
|  |  |  |  ├> "00"
|  |  |  |  |  ├> "0"
|  |  |  |  |  |  ├> ""
|  |  |  |  |  |  ├> " "
|  |  |  |  |  |  |  └> ""
|  |  |  |  |  |  ├> "("
|  |  |  |  |  |  |  ├> ""
|  |  |  |  |  |  |  ├> "$"
|  |  |  |  |  |  |  |  ├> ""
|  |  |  |  |  |  |  |  ├> "\\""
|  |  |  |  |  |  |  |  |  ├> ""
|  |  |  |  |  |  |  |  |  └> "!"
|  |  |  |  |  |  |  |  |     ├> ""
|  |  |  |  |  |  |  |  |     └> " "
|  |  |  |  |  |  |  |  |        └> ""
|  |  |  |  |  |  |  |  └> "#"
|  |  |  |  |  |  |  |     ├> ""
|  |  |  |  |  |  |  |     └> "\\""
|  |  |  |  |  |  |  |        ├> ""
|  |  |  |  |  |  |  |        ├> " "
|  |  |  |  |  |  |  |        |  └> ""
|  |  |  |  |  |  |  |        └> "!"
|  |  |  |  |  |  |  |           ├> ""
|  |  |  |  |  |  |  |           └> " "
|  |  |  |  |  |  |  |              └> ""
|  |  |  |  |  |  |  ├> "&"
|  |  |  |  |  |  |  |  ├> ""
|  |  |  |  |  |  |  |  └> "%"
|  |  |  |  |  |  |  |     ├> ""
|  |  |  |  |  |  |  |     └> "$"
|  |  |  |  |  |  |  |        ├> ""
|  |  |  |  |  |  |  |        ├> " "
|  |  |  |  |  |  |  |        |  └> ""
|  |  |  |  |  |  |  |        ├> "\\""
|  |  |  |  |  |  |  |        |  ├> ""
|  |  |  |  |  |  |  |        |  └> "!"
|  |  |  |  |  |  |  |        |     ├> ""
|  |  |  |  |  |  |  |        |     └> " "
|  |  |  |  |  |  |  |        |        └> ""
|  |  |  |  |  |  |  |        └> "#"
|  |  |  |  |  |  |  |           ├> ""
|  |  |  |  |  |  |  |           └> "\\""
|  |  |  |  |  |  |  |              ├> ""
|  |  |  |  |  |  |  |              ├> " "
|  |  |  |  |  |  |  |              |  └> ""
|  |  |  |  |  |  |  |              └> "!"
|  |  |  |  |  |  |  |                 ├> ""
|  |  |  |  |  |  |  |                 └> " "
|  |  |  |  |  |  |  |                    └> ""
|  |  |  |  |  |  |  └> "'"
|  |  |  |  |  |  |     ├> ""
|  |  |  |  |  |  |     └> "&"
|  |  |  |  |  |  |        ├> ""
|  |  |  |  |  |  |        ├> " "
|  |  |  |  |  |  |        |  └> ""
|  |  |  |  |  |  |        ├> "#"
|  |  |  |  |  |  |        |  ├> ""
|  |  |  |  |  |  |        |  └> "\\""
|  |  |  |  |  |  |        |     ├> ""
|  |  |  |  |  |  |        |     └> "!"
|  |  |  |  |  |  |        |        ├> ""
|  |  |  |  |  |  |        |        └> " "
|  |  |  |  |  |  |        |           └> ""
|  |  |  |  |  |  |        └> "%"
|  |  |  |  |  |  |           ├> ""
|  |  |  |  |  |  |           └> "$"
|  |  |  |  |  |  |              ├> ""
|  |  |  |  |  |  |              └> "#"
|  |  |  |  |  |  |                 ├> ""
|  |  |  |  |  |  |                 ├> " "
|  |  |  |  |  |  |                 |  └> ""
|  |  |  |  |  |  |                 └> "\\""
|  |  |  |  |  |  |                    ├> ""
|  |  |  |  |  |  |                    └> "!"
|  |  |  |  |  |  |                       ├> ""
|  |  |  |  |  |  |                       └> " "
|  |  |  |  |  |  |                          └> ""
|  |  |  |  |  |  ├> ","
|  |  |  |  |  |  |  ├> ""
|  |  |  |  |  |  |  ├> "*"
|  |  |  |  |  |  |  |  ├> ""
|  |  |  |  |  |  |  |  └> ")"
|  |  |  |  |  |  |  |     ├> ""
|  |  |  |  |  |  |  |     └> "("
|  |  |  |  |  |  |  |        ├> ""
|  |  |  |  |  |  |  |        ├> " "
|  |  |  |  |  |  |  |        |  └> ""
|  |  |  |  |  |  |  |        ├> "$"
|  |  |  |  |  |  |  |        |  ├> ""
|  |  |  |  |  |  |  |        |  ├> "\\""
|  |  |  |  |  |  |  |        |  |  ├> ""
|  |  |  |  |  |  |  |        |  |  └> "!"
|  |  |  |  |  |  |  |        |  |     ├> ""
|  |  |  |  |  |  |  |        |  |     └> " "
|  |  |  |  |  |  |  |        |  |        └> …
|  |  |  |  |  |  |  |        |  └> …
|  |  |  |  |  |  |  |        └> …
|  |  |  |  |  |  |  └> …
|  |  |  |  |  |  └> …
|  |  |  |  |  └> …
|  |  |  |  └> …
|  |  |  └> …
|  |  └> …
|  └> …
└> …"
`;

exports[`string (integration) > should be able to shrink 'Hey!' with fc.string({undefined}) 1`] = `
""Hey!"
├> ""
├> "y!"
|  ├> "!"
|  |  ├> ""
|  |  └> " "
|  |     └> ""
|  ├> " !"
|  |  ├> "!"
|  |  |  ├> ""
|  |  |  └> " "
|  |  |     └> ""
|  |  ├> " "
|  |  |  └> ""
|  |  └> "  "
|  |     ├> " "
|  |     |  └> ""
|  |     └> " "
|  |        └> ""
|  ├> "M!"
|  |  ├> "!"
|  |  |  ├> ""
|  |  |  └> " "
|  |  |     └> ""
|  |  ├> "7!"
|  |  |  ├> "!"
|  |  |  |  ├> ""
|  |  |  |  └> " "
|  |  |  |     └> ""
|  |  |  ├> ",!"
|  |  |  |  ├> "!"
|  |  |  |  |  ├> ""
|  |  |  |  |  └> " "
|  |  |  |  |     └> ""
|  |  |  |  ├> "&!"
|  |  |  |  |  ├> "!"
|  |  |  |  |  |  ├> ""
|  |  |  |  |  |  └> " "
|  |  |  |  |  |     └> ""
|  |  |  |  |  ├> "#!"
|  |  |  |  |  |  ├> "!"
|  |  |  |  |  |  |  ├> ""
|  |  |  |  |  |  |  └> " "
|  |  |  |  |  |  |     └> ""
|  |  |  |  |  |  ├> "\\"!"
|  |  |  |  |  |  |  ├> "!"
|  |  |  |  |  |  |  |  ├> ""
|  |  |  |  |  |  |  |  └> " "
|  |  |  |  |  |  |  |     └> ""
|  |  |  |  |  |  |  ├> "!!"
|  |  |  |  |  |  |  |  ├> "!"
|  |  |  |  |  |  |  |  |  ├> ""
|  |  |  |  |  |  |  |  |  └> " "
|  |  |  |  |  |  |  |  |     └> ""
|  |  |  |  |  |  |  |  ├> " !"
|  |  |  |  |  |  |  |  |  ├> "!"
|  |  |  |  |  |  |  |  |  |  ├> ""
|  |  |  |  |  |  |  |  |  |  └> " "
|  |  |  |  |  |  |  |  |  |     └> ""
|  |  |  |  |  |  |  |  |  ├> " "
|  |  |  |  |  |  |  |  |  |  └> ""
|  |  |  |  |  |  |  |  |  └> "  "
|  |  |  |  |  |  |  |  |     ├> " "
|  |  |  |  |  |  |  |  |     |  └> ""
|  |  |  |  |  |  |  |  |     └> " "
|  |  |  |  |  |  |  |  |        └> ""
|  |  |  |  |  |  |  |  ├> "!"
|  |  |  |  |  |  |  |  |  ├> ""
|  |  |  |  |  |  |  |  |  └> " "
|  |  |  |  |  |  |  |  |     └> ""
|  |  |  |  |  |  |  |  └> "! "
|  |  |  |  |  |  |  |     ├> " "
|  |  |  |  |  |  |  |     |  └> ""
|  |  |  |  |  |  |  |     ├> "  "
|  |  |  |  |  |  |  |     |  ├> " "
|  |  |  |  |  |  |  |     |  |  └> ""
|  |  |  |  |  |  |  |     |  └> " "
|  |  |  |  |  |  |  |     |     └> ""
|  |  |  |  |  |  |  |     └> "!"
|  |  |  |  |  |  |  |        ├> ""
|  |  |  |  |  |  |  |        └> " "
|  |  |  |  |  |  |  |           └> ""
|  |  |  |  |  |  |  ├> "\\""
|  |  |  |  |  |  |  |  ├> ""
|  |  |  |  |  |  |  |  └> "!"
|  |  |  |  |  |  |  |     ├> ""
|  |  |  |  |  |  |  |     └> " "
|  |  |  |  |  |  |  |        └> ""
|  |  |  |  |  |  |  └> "\\" "
|  |  |  |  |  |  |     ├> " "
|  |  |  |  |  |  |     |  └> ""
|  |  |  |  |  |  |     ├> "! "
|  |  |  |  |  |  |     |  ├> " "
|  |  |  |  |  |  |     |  |  └> ""
|  |  |  |  |  |  |     |  ├> "  "
|  |  |  |  |  |  |     |  |  ├> " "
|  |  |  |  |  |  |     |  |  |  └> ""
|  |  |  |  |  |  |     |  |  └> " "
|  |  |  |  |  |  |     |  |     └> ""
|  |  |  |  |  |  |     |  └> "!"
|  |  |  |  |  |  |     |     └> …
|  |  |  |  |  |  |     └> …
|  |  |  |  |  |  └> …
|  |  |  |  |  └> …
|  |  |  |  └> …
|  |  |  └> …
|  |  └> …
|  └> …
└> …"
`;

exports[`string (integration) > should be able to shrink 'abc!123!' with fc.string({PatternsArbitrary{ patterns: [ …(3) ] }}) 1`] = `
""abc!123!"
├> ""
├> "123!"
|  ├> "!"
|  |  ├> ""
|  |  └> "123"
|  |     └> ""
|  ├> "123"
|  |  └> ""
|  └> "123123"
|     ├> "123"
|     |  └> ""
|     └> "123"
|        └> ""
├> "!123!"
|  ├> "123!"
|  |  ├> "!"
|  |  |  ├> ""
|  |  |  └> "123"
|  |  |     └> ""
|  |  ├> "123"
|  |  |  └> ""
|  |  └> "123123"
|  |     ├> "123"
|  |     |  └> ""
|  |     └> "123"
|  |        └> ""
|  ├> "123123!"
|  |  ├> "123!"
|  |  |  ├> "!"
|  |  |  |  ├> ""
|  |  |  |  └> "123"
|  |  |  |     └> ""
|  |  |  ├> "123"
|  |  |  |  └> ""
|  |  |  └> "123123"
|  |  |     ├> "123"
|  |  |     |  └> ""
|  |  |     └> "123"
|  |  |        └> ""
|  |  ├> "123"
|  |  |  └> ""
|  |  ├> "123!"
|  |  |  ├> "!"
|  |  |  |  ├> ""
|  |  |  |  └> "123"
|  |  |  |     └> ""
|  |  |  ├> "123"
|  |  |  |  └> ""
|  |  |  └> "123123"
|  |  |     ├> "123"
|  |  |     |  └> ""
|  |  |     └> "123"
|  |  |        └> ""
|  |  ├> "123123"
|  |  |  ├> "123"
|  |  |  |  └> ""
|  |  |  └> "123"
|  |  |     └> ""
|  |  └> "123123123"
|  |     ├> "123123"
|  |     |  ├> "123"
|  |     |  |  └> ""
|  |     |  └> "123"
|  |     |     └> ""
|  |     ├> "123"
|  |     |  └> ""
|  |     ├> "123123"
|  |     |  ├> "123"
|  |     |  |  └> ""
|  |     |  └> "123"
|  |     |     └> ""
|  |     └> "123123"
|  |        ├> "123"
|  |        |  └> ""
|  |        └> "123"
|  |           └> ""
|  ├> "!"
|  |  ├> ""
|  |  └> "123"
|  |     └> ""
|  ├> "!!"
|  |  ├> "!"
|  |  |  ├> ""
|  |  |  └> "123"
|  |  |     └> ""
|  |  ├> "123!"
|  |  |  ├> "!"
|  |  |  |  ├> ""
|  |  |  |  └> "123"
|  |  |  |     └> ""
|  |  |  ├> "123"
|  |  |  |  └> ""
|  |  |  └> "123123"
|  |  |     ├> "123"
|  |  |     |  └> ""
|  |  |     └> "123"
|  |  |        └> ""
|  |  ├> "!"
|  |  |  ├> ""
|  |  |  └> …
|  |  └> …
|  └> …
└> …"
`;

exports[`string (integration) > should be able to shrink 'abc!123!' with fc.string({undefined}) 1`] = `
""abc!123!"
├> ""
├> "123!"
|  ├> "3!"
|  |  ├> "!"
|  |  |  ├> ""
|  |  |  └> " "
|  |  |     └> ""
|  |  ├> " !"
|  |  |  ├> "!"
|  |  |  |  ├> ""
|  |  |  |  └> " "
|  |  |  |     └> ""
|  |  |  ├> " "
|  |  |  |  └> ""
|  |  |  └> "  "
|  |  |     ├> " "
|  |  |     |  └> ""
|  |  |     └> " "
|  |  |        └> ""
|  |  ├> "*!"
|  |  |  ├> "!"
|  |  |  |  ├> ""
|  |  |  |  └> " "
|  |  |  |     └> ""
|  |  |  ├> "%!"
|  |  |  |  ├> "!"
|  |  |  |  |  ├> ""
|  |  |  |  |  └> " "
|  |  |  |  |     └> ""
|  |  |  |  ├> "#!"
|  |  |  |  |  ├> "!"
|  |  |  |  |  |  ├> ""
|  |  |  |  |  |  └> " "
|  |  |  |  |  |     └> ""
|  |  |  |  |  ├> "\\"!"
|  |  |  |  |  |  ├> "!"
|  |  |  |  |  |  |  ├> ""
|  |  |  |  |  |  |  └> " "
|  |  |  |  |  |  |     └> ""
|  |  |  |  |  |  ├> "!!"
|  |  |  |  |  |  |  ├> "!"
|  |  |  |  |  |  |  |  ├> ""
|  |  |  |  |  |  |  |  └> " "
|  |  |  |  |  |  |  |     └> ""
|  |  |  |  |  |  |  ├> " !"
|  |  |  |  |  |  |  |  ├> "!"
|  |  |  |  |  |  |  |  |  ├> ""
|  |  |  |  |  |  |  |  |  └> " "
|  |  |  |  |  |  |  |  |     └> ""
|  |  |  |  |  |  |  |  ├> " "
|  |  |  |  |  |  |  |  |  └> ""
|  |  |  |  |  |  |  |  └> "  "
|  |  |  |  |  |  |  |     ├> " "
|  |  |  |  |  |  |  |     |  └> ""
|  |  |  |  |  |  |  |     └> " "
|  |  |  |  |  |  |  |        └> ""
|  |  |  |  |  |  |  ├> "!"
|  |  |  |  |  |  |  |  ├> ""
|  |  |  |  |  |  |  |  └> " "
|  |  |  |  |  |  |  |     └> ""
|  |  |  |  |  |  |  └> "! "
|  |  |  |  |  |  |     ├> " "
|  |  |  |  |  |  |     |  └> ""
|  |  |  |  |  |  |     ├> "  "
|  |  |  |  |  |  |     |  ├> " "
|  |  |  |  |  |  |     |  |  └> ""
|  |  |  |  |  |  |     |  └> " "
|  |  |  |  |  |  |     |     └> ""
|  |  |  |  |  |  |     └> "!"
|  |  |  |  |  |  |        ├> ""
|  |  |  |  |  |  |        └> " "
|  |  |  |  |  |  |           └> ""
|  |  |  |  |  |  ├> "\\""
|  |  |  |  |  |  |  ├> ""
|  |  |  |  |  |  |  └> "!"
|  |  |  |  |  |  |     ├> ""
|  |  |  |  |  |  |     └> " "
|  |  |  |  |  |  |        └> ""
|  |  |  |  |  |  └> "\\" "
|  |  |  |  |  |     ├> " "
|  |  |  |  |  |     |  └> ""
|  |  |  |  |  |     ├> "! "
|  |  |  |  |  |     |  ├> " "
|  |  |  |  |  |     |  |  └> ""
|  |  |  |  |  |     |  ├> "  "
|  |  |  |  |  |     |  |  ├> " "
|  |  |  |  |  |     |  |  |  └> ""
|  |  |  |  |  |     |  |  └> " "
|  |  |  |  |  |     |  |     └> ""
|  |  |  |  |  |     |  └> "!"
|  |  |  |  |  |     |     ├> ""
|  |  |  |  |  |     |     └> " "
|  |  |  |  |  |     |        └> ""
|  |  |  |  |  |     └> "\\""
|  |  |  |  |  |        ├> ""
|  |  |  |  |  |        └> "!"
|  |  |  |  |  |           ├> ""
|  |  |  |  |  |           └> " "
|  |  |  |  |  |              └> ""
|  |  |  |  |  └> …
|  |  |  |  └> …
|  |  |  └> …
|  |  └> …
|  └> …
└> …"
`;
